### PR TITLE
[GAMM] Add test for `GetMaximalNoSwapLPAmount`

### PR DESCRIPTION
**Pull request by ThanhNhann**
Closes: #3120 

## Brief Changelog

Refactor `GetMaximalNoSwapLPAmount` function
Write test cases for it

## Testing and Verifying

- Tested by test cases

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not documented)

### CODE REVIEW

File: x/gamm/pool-models/stableswap/pool.go [hunk 1]
```diff
@@ -416,6 +416,36 @@ func (p *Pool) SetScalingFactors(ctx sdk.Context, scalingFactors []uint64, sende
 	return nil
 }
 
+// GetMaximalNoSwapLPAmount returns the coins(lp liquidity) needed to get the specified amount of shares in the pool.
+// Steps to getting the needed lp liquidity coins needed for the share of the pools are
+// 1. calculate how much percent of the pool does given share account for(# of input shares / # of current total shares)
+// 2. since we know how much % of the pool we want, iterate through all pool liquidity to calculate how much coins we need for
+// each pool asset.
+func (p Pool) GetMaximalNoSwapLPAmount(ctx sdk.Context, shareOutAmount sdk.Int) (neededLpLiquidity sdk.Coins, err error) {
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/pool.go:424]: Why are we duplicating the implementations? If both balancer and stableswap need to have essentially the same logic, keeping it as function makes sense to me

**Review by pysel — COMMENTED** [review]
thanks for the PR! I personally agree with @p0mvn about leaving the function unexported. Also left some of my personal comments

**Comment by ThanhNhann** [comment]
thank @p0mvn and @pysel for review, the pr was updated

**Comment by mattverse** [comment]
Changing PR name seeing no edits in the logic itself

File: x/gamm/keeper/pool_test.go [hunk 2]
```diff
@@ -517,3 +517,152 @@ func (suite *KeeperTestSuite) TestSetStableSwapScalingFactors() {
...
+
+		"Balancer pool: Pass": {
+			poolId: suite.prepareCustomBalancerPool(defaultAcctFunds, defaultPoolAssets, balancer.PoolParams{
+				SwapFee: defaultSwapFee,
+				ExitFee: defaultZeroExitFee,
+			}),
+			shareOutAmount: sdk.NewInt(8_000_000_000_000_000_000),
+			expectedLpLiquidity: sdk.Coins{
+				sdk.NewInt64Coin(defaultPoolAssets[1].Token.Denom, 800),
+				sdk.NewInt64Coin(defaultPoolAssets[0].Token.Denom, 800),
```
Comment (mattverse) [inline x/gamm/keeper/pool_test.go:554]: Can we add one line comments for how we are getting these numbers?

File: x/gamm/keeper/pool_test.go [hunk 3]
```diff
@@ -517,3 +517,152 @@ func (suite *KeeperTestSuite) TestSetStableSwapScalingFactors() {
...
+			},
+		},
```
Comment (mattverse) [inline x/gamm/keeper/pool_test.go:556]: Can we add test cases that handles the ceiling / rounding numbers?

**Comment by mattverse** [comment]
@ThanhNhann Can you re-assign me for review once r4r?

**Review by mattverse — APPROVED** [review]
LGTM

**Comment by czarcas7ic** [comment]
I had deleted the no-changelog label in favor of the A:no-changelog label, I don't think this PR needs a changelog entry, I will remove it

File: CHANGELOG.md [hunk 4]
```diff
@@ -52,6 +52,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Misc Improvements
 
+  * [#5161] (https://github.com/osmosis-labs/osmosis/pull/5161) Add test for GetMaximalNoSwapLPAmount
```
Comment (czarcas7ic) [inline CHANGELOG.md:55]:
```suggestion
```

Comment (czarcas7ic) [reply CHANGELOG.md:55]: actually @ThanhNhann I am unable to make the change since this isn't coming from our repo. Can you make this change?

**Review by czarcas7ic — APPROVED** [review]

**Review by pysel — APPROVED** [review]

File: CHANGELOG.md (hunk 4 above)
Comment (ThanhNhann) [reply CHANGELOG.md]: Updated @czarcas7ic

**Comment by ThanhNhann** [comment]
hey @czarcas7ic, everything is good now. Can you merge it?

**Comment by czarcas7ic** [comment]
@ThanhNhann merged, thanks!

**Pull request merged**